numcodecs/zfpy.py

```diff
@@ -0,0 +1,90 @@
+_zfpy = None
+try:
+    import zfpy as _zfpy
+except ImportError:  # pragma: no cover
+    pass
+
+
+if _zfpy and sys.version_info < (3, 8):
+
+    from .abc import Codec
+    from .compat import ndarray_copy, ensure_contiguous_ndarray, ensure_bytes
+
+    # noinspection PyShadowingBuiltins
+    class ZFPY(Codec):
+        """Codec providing compression using zfpy via the Python standard
+        library.
+
+        Parameters
+        ----------
+        mode : integer
+            One of the zfpy mode choice, e.g., ``zfpy.mode_fixed_accuracy``.
+        tolerance : double, optional
+            A double-precision number, specifying the compression accuracy needed.
+        rate : double, optional
+            A double-precision number, specifying the compression rate needed.
+        precision : int, optional
+            A integer number, specifying the compression precision needed.
+
+        """
+
+        codec_id = 'zfpy'
+
+        def __init__(self, mode=_zfpy.mode_fixed_accuracy, tolerance=-1,
+                     rate=-1, precision=-1, compression_kwargs=None):
+            self.mode = mode
+            if mode == _zfpy.mode_fixed_accuracy:
+                self.compression_kwargs = {
+                    "tolerance": tolerance
+                }
+            elif mode == _zfpy.mode_fixed_rate:
+                self.compression_kwargs = {
+                    "rate": rate
+                }
+            elif mode == _zfpy.mode_fixed_precision:
+                self.compression_kwargs = {
+                    "precision": precision
+                }
+            else:
+                pass
+
+            self.tolerance = tolerance
+            self.rate = rate
+            self.precision = precision
+
+        def encode(self, buf):
+
+            # normalise inputs
+            buf = ensure_contiguous_ndarray(buf)
+
+            # do compression
+            comp_arr = _zfpy.compress_numpy(buf, write_header=True, **self.compression_kwargs)
+
+            return comp_arr
+
+        def decode(self, buf, out=None):
+
+            # normalise inputs
+            buf = ensure_bytes(buf)
+            if out is not None:
+                out = ensure_contiguous_ndarray(out)
+
+            # do decompression
+            dec = _zfpy.decompress_numpy(buf)
```

Comment:
Guessing by the name this requires a NumPy array. Currently we are forcing `buf` to a `bytes` object. Would update the line discussed above to `ensure_ndarray` to fix that.

Reply:
This comment does not seem resolved to me.